git/filter_protocol.go

```diff
@@ -0,0 +1,114 @@
+package git
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"strconv"
+	"strings"
+)
+
+type protocol struct {
+	r *bufio.Reader
+	w *bufio.Writer
+}
+
+func newProtocolRW(r io.Reader, w io.Writer) *protocol {
+	return &protocol{
+		r: bufio.NewReader(r),
+		w: bufio.NewWriter(w),
+	}
+}
+
+func (p *protocol) readPacket() ([]byte, error) {
+	pktLenHex, err := ioutil.ReadAll(io.LimitReader(p.r, 4))
+	if err != nil || len(pktLenHex) != 4 { // TODO check pktLenHex length
+		return nil, err
+	}
+
+	pktLen, err := strconv.ParseInt(string(pktLenHex), 16, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	if pktLen == 0 {
+		return nil, nil
+	}
+	if pktLen <= 4 {
+		return nil, errors.New("Invalid packet length.")
+	}
+
+	return ioutil.ReadAll(io.LimitReader(p.r, pktLen-4))
+}
+
+func (p *protocol) readPacketText() (string, error) {
+	data, err := p.readPacket()
+	return strings.TrimSuffix(string(data), "\n"), err
+}
+
+func (p *protocol) readPacketList() ([]string, error) {
+	var list []string
+	for {
+		data, err := p.readPacketText()
+		if err != nil {
+			return nil, err
+		}
+
+		if len(data) == 0 {
+			break
```

Comment:
E.g. here we want to wait for a flush. If there is "no" package then we want to read from the pope, again.